packages/upload-api/src/blob/add.js

```diff
@@ -0,0 +1,376 @@
+ * @param {API.BlobModel} scheduleAllocateProps.blob
+ * @param {API.IssuedInvocationView<API.BlobAllocate>} scheduleAllocateProps.allocate
+ * @param {API.Invocation<API.BlobAllocate>} scheduleAllocateProps.allocatefx
+ */
+async function scheduleAllocate({ context, blob, allocate, allocatefx }) {
+  /** @type {import('@ucanto/interface').Receipt<import('@web3-storage/capabilities/types').BlobAllocateSuccess> | undefined} */
+  let blobAllocateReceipt
+
+  // Get receipt for `blob/allocate` if available, otherwise schedule invocation
+  const receiptGet = await context.receiptsStorage.get(allocatefx.link())
+  if (receiptGet.error && receiptGet.error.name !== 'RecordNotFound') {
+    return {
+      error: receiptGet.error,
+    }
+  } else if (receiptGet.ok) {
+    // @ts-expect-error ts not able to cast receipt
+    blobAllocateReceipt = receiptGet.ok
+
+    // Verify if allocation is expired before "accepting" this receipt.
+    // Note that if there is no address, means it was already allocated successfully before
+    const expiresAt = blobAllocateReceipt?.out.ok?.address?.expiresAt
+    if (expiresAt && (new Date()).getTime() > (new Date(expiresAt)).getTime()) {
+      // if expired, we must see if blob was written to avoid allocating one more time
+      const hasBlobStore = await context.blobsStorage.has(blob.digest)
+      if (hasBlobStore.error) {
+        return hasBlobStore
+      } else if (!hasBlobStore.ok) {
+        blobAllocateReceipt = undefined
+      }
+    }
+  }
+
+  // if not already allocated (or expired) schedule `blob/allocate`
+  if (!blobAllocateReceipt) {
+    // Execute allocate invocation
+    const allocateRes = await allocate.execute(context.getServiceConnection())
+    if (allocateRes.out.error) {
+      return {
+        error: new AwaitError({
+          cause: allocateRes.out.error,
+          at: 'ucan/wait',
+          reference: ['.out.ok', allocatefx.cid],
+        }),
+      }
+    }
+    blobAllocateReceipt = allocateRes
+  }
+
+  // Create `blob/allocate` receipt as conclude invocation to inline as effect
+  const concludeAllocate = createConcludeInvocation(
+    context.id,
+    context.id,
+    blobAllocateReceipt
+  )
+  return {
+    ok: {
+      concludeAllocatefx: await concludeAllocate.delegate(),
+    },
+  }
+}
+
+/**
+ * Create `blob/add` next tasks.
+ *
+ * @param {object} nextProps
+ * @param {API.BlobServiceContext} nextProps.context
+ * @param {API.BlobModel} nextProps.blob
+ * @param {API.DIDKey} nextProps.space
+ * @param {API.Link} nextProps.cause
+ */
+async function createNextTasks({ context, blob, space, cause }) {
+  // 1. Create web3.storage/blob/allocate invocation and task
+  const allocate = W3sBlob.allocate.invoke({
+    issuer: context.id,
+    audience: context.id,
+    with: context.id.did(),
+    nb: {
+      blob,
+      cause: cause,
+      space,
+    },
+    expiration: Infinity,
+  })
+  const allocatefx = await allocate.delegate()
```

Comment:
I don't think you need both `allocate` and `allocate.delegate()` because you could do `connection.execute(allocatefx)` instead of `allocate.execute(connection)`.

Reply:
I need the delegate to have a CID for it, the `allocate` does not give it to me